docs(enhancements): replace exceptions.yaml with Exception Requests Project view board (#3023)
kubernetes/sig-release: release-team/role-handbooks/enhancements/README.md

Reviewer: Historically, we’ve had feedback that the Enhancements `README.md` is getting large. Could we move the exception section into a separate file and link it here, similar to the communication templates?

_Just a suggestion not a merge blocker, it’s okay if we don’t do it in this PR._

Author: Sure, I will quickly raise another PR to move it to a separate file after this PR is merged.

Author: Created a todo item: https://github.com/kubernetes/sig-release/issues/3039

Ensure that the previous Enhancements Lead has given you access to:

> - The previous Kubernetes Release Tracking Board.

Reviewer 2: some of these renames will conflict with https://github.com/kubernetes/sig-release/pull/3022

### Exceptions

The exception process (criteria, deadlines, and what enhancement owners need to submit) is outlined in [releases/EXCEPTIONS.md](/releases/EXCEPTIONS.md).

> Starting with the v1.36 release cycle, the Release Team tracks all exception requests on the [Enhancements Tracking Board](#working-with-the-enhancement-tracking-board) using the `Exception Requests` view (e.g., [v1.36 Exception Requests view](https://github.com/orgs/kubernetes/projects/241/views/7)), instead of maintaining an `exceptions.yaml` file in the release directory.

Reviewer 3: Now that we are going with the new name for `Enhancements Tracking Board` to `Release Tracking Board` should we also replace it here?

Author: Oh that's correct. Will update it.

#### Fields on the Exception Requests view

The `Exception Requests` view filters on `has:exception-request-type` and surfaces the following fields for each request:

|                          Field | Description                                                                                                                  |
|-------------------------------:|------------------------------------------------------------------------------------------------------------------------------|
|                          Title | Title of (and link to) the KEP `Issue` in k/enhancements. For additional exceptions on the same KEP, this is a draft item.   |
|                         Status | The KEP's current tracking status (e.g. `Tracked for PRR freeze`, `Removed from Milestone`).                                 |
|                            SIG | Owning SIG.                                                                                                                  |
|         Exception Request Type | The freeze the exception applies to: `PRR Freeze`, `Enhancements Freeze`, `Code and Test Freeze`, or `Docs Freeze`.          |

> |               Exception Status | `Approved` or `Rejected` once the Release Team has made a decision.                                                          |

Reviewer 2: I can't currently see the board details, but I recall we also had other statuses such as "under review" i think? If so, we should mention them here too

Author: Oh yeah, we have In Review as well. adding it.

|         Exception Request Link | Link to the Google Groups thread where the request was filed.                                                                |

> |  Exception Request Slack Thread | Link to the corresponding `#release-enhancements` Slack thread.                                                              |

Reviewer 2: ```suggestion
|  Exception Request Slack Thread | Link to the corresponding `#sig-release` Slack thread.                                                              |
```

|          Exception Related PRs | Links to PRs associated with the exception (k/enhancements PRs for PRR/Enhancements Freeze, k/k PRs for Code/Test Freeze).   |
|                Additional Time | The additional time requested by the enhancement owner, in calendar days.                                                    |
|         Exception Request Date | Date the exception was requested.                                                                                            |

> |        Exception Decision Date | Date the Release Team approved or rejected the exception.                                                                    |

Reviewer 2: maybe mention that both should be in UTC time, just to keep them standardized

> |                Exception Lead | The Enhancements Lead shadow or Release Lead shadow assigned to follow up on the request.                                    |

Reviewer 2: related to my other comment in EXCEPTIONS.md - we have previously rotated between Enhancements Lead, Release Lead, Release Lead Shadows. Although maybe it's a good idea to get Enhancements Shadows in the rotation too? If so we should keep the wording between here and EXCEPTIONS.md in sync

Author: umm yeah I think it's a point for discussion. I will add an open item in the next release team call.

Reviewer: In my mind, this should stay as a Release Lead / Lead Shadows and Enhancements Lead activity. The leads will need to kick-off Slack thread to evaluate exception requests, so it makes sense for them to do this (similar to how this worked previously with the exception tracking spreadsheet).

Reviewer: We should also link to the exception tracking steps from the Release Lead / Lead Shadows task handbook in a **separate PR**, likely somewhere near the [Slack thread section](https://github.com/kubernetes/sig-release/blob/master/release-team/role-handbooks/release-team-lead/README.md#code-freeze). That would also help with the retro feedback around clearer Lead Shadow tasks and responsibilities.

Author: Sure Dipesh, makes sense. Will add this in the other PR that I will raise to separate the exception section.

> |                Exception Lead | The Enhancements Lead shadow or Release Lead shadow assigned to follow up on the request.                                    |

Reviewer 2: ```suggestion
|                Exception Liaison | The Enhancements Lead shadow or Release Lead shadow assigned to follow up on the request.                                    |
```
in the 1.36 board it was called `Exception Liaison`

> 

Reviewer 2: there's also another column `Exception Request Notes` for any additional notes about the KEP / exception

Author: Oh right I missed it earlier. Added it now.

#### Adding a new exception request

When an exception request email comes in:

> 1. **Locate the KEP in the `Enhancements` view** by searching for the KEP number.

Reviewer 2: we should also mention updating the `Status` on the `Enhancements` view of the board:
- when the exception first comes in, we should set it to `Exception Pending`
- once the Release Team has decided to approve the exception, set it to `Exception Approved` (or if rejected, set it to `Removed from Milestone`)
- once the KEP has satisfied all the requirements, move it to `Tracked for...` (or if they missed the new deadline, set it to `Removed from Milestone`)

whenever removing from milestone, we should also clear the tracked/milestone labels:
```
/remove-label tracked/yes
/label tracked/no
/milestone clear
```

when adding it back to the milestone as `Tracked...`, we should be sure to update the labels as well:
```
/label tracked/yes
/remove-label tracked/no
/milestone {milestone}
```

Author: Makes sense, added it in steps.

    - If the KEP does not appear, it is likely being filtered out by the default view filter (e.g., `-status:Deferred,"Removed from Milestone" -enhancement-type:Docs`). Temporarily remove the relevant filter so the KEP becomes visible.
2. **Set the `Exception Request Type`** field on the KEP to the appropriate freeze (`PRR Freeze`, `Enhancements Freeze`, `Code and Test Freeze`, or `Docs Freeze`).
3. As soon as `Exception Request Type` has a value, the KEP automatically appears in the `Exception Requests` view. Switch to that view and fill in the remaining fields (`Exception Request Link`, `Exception Request Slack Thread`, `Exception Related PRs`, `Additional Time`, `Exception Request Date`, etc.).

> 4. Assign the request to the Enhancements Lead or an Release Lead shadow via the `Exception Lead` field for follow-up on the Slack thread.

Reviewer 2: ```suggestion
4. Assign the request to the Enhancements Lead or a Release Lead shadow via the `Exception Liaison` field for follow-up on the Slack thread.
```

5. Once the Release Team has decided on the request, set the `Exception Status` to `Approved` or `Rejected` and populate `Exception Decision Date`.

#### Handling multiple exceptions for the same KEP

> GitHub Projects does not support multi-select on the `Exception Request Type` field, so a single KEP issue can only represent one exception type at a time. If the same KEP requests exceptions for multiple freezes, the KEP will already appear in the `Exception Requests` view (with the original freeze type) and cannot be re-added from the same issue.

Reviewer 3: the new `multiple exceptions` workflow conflicts with the existing policy in `releases/EXCEPTIONS.md`, which tells enhancement owners that multiple exceptions will not be granted. Should we either scope the handbook text to the already-allowed combined PRR + Enhancements exception case, or update the policy to explicitly define when multiple exception entries are acceptable?

Author: Good catch, thanks for flagging this. `releases/EXCEPTIONS.md` does touch on multiple exceptions in two places:

- The note at the end of the "Enhancement Owners" section: "we will not grant multiple exceptions for a enhancement".
- The PRR-freeze carve-out that lets owners file a combined PRR + Enhancements Freeze request.

In `release-team/role-handbooks/enhancements/README.md` I am mentioning in case the same KEP legitimately needs separate exceptions at different freezes across the cycle (e.g. a PRR Freeze exception, later on, a Code/Test Freeze exception). That has happened in past releases. 
The "we will not grant multiple exceptions" line as I read it is about not re-granting/extending an exception for the same freeze (e.g. asking for 3 more days after a 3-day exception was already approved), not about prohibiting separate exceptions at PRR, Enhancements, Code/Test, and Docs Freeze across a release.

I think the cleanest fix is to update `releases/EXCEPTIONS.md` to make that distinction explicit, since it's currently ambiguous. Something like:

```
The Release Team will not grant multiple exceptions for the same freeze for a single enhancement (i.e. once an exception has been approved or rejected for a given freeze, the decision is final and additional time cannot be requested). A KEP may still request separate exceptions at different freeze points across the release cycle (e.g. PRR Freeze, Enhancements Freeze and later Code/Test Freeze), each evaluated on its own merits.
```
Let me know if this looks good. 
@dipesh-rawat @jenshu

Reviewer: I agree with Rayan. The current wording in releases/EXCEPTIONS.md is unclear and worth cleaning up. Could you commit the proposed change?

Author: I've included it and updated the PR.

### Exceptions

The exception process (criteria, deadlines, and what enhancement owners need to submit) is outlined in [releases/EXCEPTIONS.md](/releases/EXCEPTIONS.md).

> Starting with the v1.36 release cycle, the Release Team tracks all exception requests on the [Release Tracking Board](#working-with-the-enhancement-tracking-board) using the `Exception Requests` view (e.g., [v1.36 Exception Requests view](https://github.com/orgs/kubernetes/projects/241/views/7)), instead of maintaining an `exceptions.yaml` file in the release directory.

Reviewer 2: ```suggestion
Starting with the v1.36 release cycle, the Release Team tracks all exception requests on the [Release Tracking Board](#working-with-the-release-tracking-board) using the `Exception Requests` view (e.g., [v1.36 Exception Requests view](https://github.com/orgs/kubernetes/projects/241/views/7)), instead of maintaining an `exceptions.yaml` file in the release directory.
```

|               Exception Status | `In Review` while the Release Team is evaluating the request, then `Approved` or `Rejected` once a decision has been made.   |
|         Exception Request Link | Link to the Google Groups thread where the request was filed.                                                                |
|  Exception Request Slack Thread | Link to the corresponding `#sig-release` Slack thread.                                                              |
|          Exception Related PRs | Links to PRs associated with the exception (k/enhancements PRs for PRR/Enhancements Freeze, k/k PRs for Code/Test Freeze).   |
|                Additional Time | The additional time requested by the enhancement owner, in calendar days.                                                    |
|         Exception Request Date | Date (UTC) the exception was requested.                                                                                            |
|        Exception Decision Date | Date (UTC) the Release Team approved or rejected the exception.                                                                    |

> |                Exception Liaison | The Enhancements Lead shadow or Release Lead shadow assigned to follow up on the request.                                    |

Reviewer 2: ```suggestion
|                Exception Liaison | The Release Team member (Enhancements Lead, Release Lead, or Release Lead shadow) assigned to follow up on the request.                                    |
```

#### Adding a new exception request

When an exception request email comes in:

1. **Locate the KEP in the `Enhancements` view** by searching for the KEP number.
    - If the KEP does not appear, it is likely being filtered out by the default view filter (e.g., `-status:Deferred,"Removed from Milestone" -enhancement-type:Docs`). Temporarily remove the relevant filter so the KEP becomes visible.
2. **Set the `Exception Request Type`** field on the KEP to the appropriate freeze (`PRR Freeze`, `Enhancements Freeze`, `Code and Test Freeze`, or `Docs Freeze`).
3. **Set the `Status` field on the `Enhancements` view to `Exception Pending`** to indicate the request is open and awaiting a Release Team decision.
4. As soon as `Exception Request Type` has a value, the KEP automatically appears in the `Exception Requests` view. Switch to that view, set `Exception Status` to `In Review`, and fill in the remaining fields (`Exception Request Link`, `Exception Request Slack Thread`, `Exception Related PRs`, `Additional Time`, `Exception Request Date`, etc.).

> 5. Assign the request to the Enhancements Lead or a Release Lead shadow via the `Exception Liaison` field for follow-up on the Slack thread.

Reviewer 2: ```suggestion
5. Assign the request to the Enhancements Lead, Release Lead, or Release Lead shadow via the `Exception Liaison` field for follow-up on the Slack thread.
```
